npe2/manifest/contributions/_contributions.py

```diff
@@ -21,11 +23,21 @@ class ContributionPoints(BaseModel):
     themes: Optional[List[ThemeContribution]]
 
     # We use a dict for menus to allow for keys with `/`
-    menus: Optional[Dict[str, List[MenuItem]]] = Field(None, hide_docs=True)
+    menus: Optional[MenuContribution] = Field(None, hide_docs=True)
     submenus: Optional[List[SubmenuContribution]] = Field(None, hide_docs=True)
 
     # configuration: Optional[JsonSchemaObject]
     # keybindings: Optional[List[KeyBindingContribution]]
 
+    @validator("menus")
+    def _check_napari_menus(cls, v):
+        for key in v.keys():
+            if key not in VALID_MENUS_NAMES:
```

Comment:
I'm also not sure if this is still going to work if we need to go back on our plan to disallow plugins to populate other plugin's menus.  Quick reminder: this came up on zulip recently, where someone wanted to be able to have, for example, their own "group" menu (e.g. `Smith Lab Plugins`) ... and populate it across their multiple plugins.

If we state that only our known menu keys are valid, that won't work

Reply:
Ah, I forgot about that. What if we just restricted the top-level menus, but allowed things to be more flexible for sub-menus?

Reply:
but how do you determine what is top level and what is not?   put another way: how do you envision plugin B contributing to plugin A's menu if not by adding their menu key to the list of menus

Reply:
i know this is complicated/tricky stuff :joy: ... if you have a vision for how to restrict/validate this _at manifest ingestion time_ without also breaking the ability to add to arbitrary plugin's menus.  Can you explain it in a bit more detail (with examples)

Reply:
> I'm also not sure if this is still going to work if we need to go back on our plan to disallow plugins to populate other plugin's menus.  Quick reminder: this came up on zulip recently, where someone wanted to be able to have, for example, their own "group" menu (e.g. Smith Lab Plugins)

ah right - i might loop @chili-chiu into this discussion too to get another opinion here too. i'm not sure what's best, but that's an understandable request

Reply:
Using the sample:
```yaml
  menus:
    /napari/layer_context:
      - submenu: my-plugin.mysubmenu
      - command: my-plugin.hello_world
  submenus:
    - id: my-plugin.mysubmenu
      label: My SubMenu
      contents:
        - command: my-plugin.another_command
```

Another plugin could do:
```yaml
  menus:
    /napari/layer_context:
      - submenu: another-plugin.mysubmenu
  submenus:
    - id: another-plugin.mysubmenu
      label: My SubMenu
      contents:
        - command: another-plugin.super_sweet_command
```

I'm imagining the result in napari would be:
```
<layer context menu>
  - My SubMenu 
     - <command title for my-plugin.another_command>
     - <command title for another-plugin.super_sweet_command>
```

Reply:
@tlambert03 could we
1) go with the original menus/submenus construction approach
2) allow plugin A to contribute to submenu defined by plugin B
3) have approved locations like `/napari/tools/classification` that populate top of menu
4) put custom submenus like `napari/tools/Smith Lab` below a divider (like VScode had those submenu section, remember?)

this would be very flexible, but at least encourage use of our organizational system. thoughts?

Reply:
for steps 3/4 ... exactly where an item appears within a menu is yet to be determined.  So, the answer is kind of "yes, nothing so far limits us from doing that".  But I bet we'll come into issues with it later.  This comes back to the idea of menu groups and ordering that we discussed at the hackathon.  Here's how vscode deals with it: https://code.visualstudio.com/api/references/contribution-points#Sorting-of-groups

Anyway, I see steps 3/4 (the "where" question) as slightly orthogonal to steps 1/2 (the "what" question).  We can revisit the "where" question later (I think?)

Reply:
> I'd propose we go back to not having submenus.contents at all

Good point.

I'll try to restructure for the vscode-like design (which I do like).  I think the basic idea is the same though: Join submenus in napari based on the label - I'm imagining that's a napari-side behavior.

So the sample plugin becomes:
```yaml
  menus:
    /napari/layer_context:
      - submenu: my-plugin.mysubmenu
      - command: my-plugin.hello_world
    my-plugin.mysubmenu:
        - command: my-plugin.another_command
  submenus:
    - id: my-plugin.mysubmenu
      label: My SubMenu
```

And the other plugin is:

```yaml
  menus:
    /napari/layer_context:
      - submenu: another-plugin.mysubmenu
    another-plugin.mysubmenu:
      - command: another-plugin.super_sweet_command
  submenus:
    - id: another-plugin.mysubmenu
      label: My SubMenu
```

Same output as before
```
<layer context menu>
  - My SubMenu 
     - <command title for my-plugin.another_command>
     - <command title for another-plugin.super_sweet_command>
```

Reply:
correct... but I think you left out the key example there, which would be...

The second plugin adds an item to the first plugin's menu.  (whether the second plugin adds a new submenu is fine, but not the critical thing we're trying to determine here)

```yaml
  menus:
    /napari/layer_context:
      - submenu: another-plugin.mysubmenu
    my-plugin.mysubmenu:
      - command: another-plugin.super_sweet_command
```

Reply:
I'm thinking a plugin only refers to its own menu ids; it's self-contained. It adds to another plugin's menu only by the coincidence of using the same label ("My SubMenu" above).

In your example, if "my-plugin.submenu" moved then "another-plugin.super_sweet_command" would move along with it. In my proposal, it wouldn't.  Does that make sense?  What's the desired behavior?

Reply:
> I'm thinking a plugin only refers to its own menu ids; it's self-contained. It adds to another plugin's menu only by the coincidence of using the same label ("My SubMenu" above).

so, you're essentially just saying that "label" _is_ the identifier?   that feels icky.  The label is how the menu _appears_, it shouldn't be the globally unique identifier.  What if two labs just happen to use the same name for their submenu, but one wants to place theirs somewhere and the other wants to place it somewhere else?  In your example, they just "happened" to both be in `/napari/layer_context:` ... but what if they weren't?  How would the second party indicate which of the two menus they wanted?  (answer: the globally unique identifier that we asked the submenu creator to create in the first place :joy:).  That one's a hard no for me :)  labels make bad identifiers.

> In your example, if "my-plugin.submenu" moved then "another-plugin.super_sweet_command" would move along with it. 

That's by design.  Remember, despite the name, a submenu doesn't "live" anywhere specific.  In fact, it can live in multiple places (and in vscode they very often do).  Don't think of it as "this submenu always lives in layers context".  Think of it as "my lab offers a menu... if you'd like to participate in it, here's our unique id".

Reply:
> > In your example, if "my-plugin.submenu" moved then "another-plugin.super_sweet_command" would move along with it.
> 
> That's by design.

Ah, I was wondering about that. If that's a requirement then my thing definitely is a no-go. 

> labels make bad identifiers

Agreed. The identifier determines how submenus get attached to different places, but when they do it can happen that there are two things with the same label. We could choose to combine them at that point. 

In any case, I've probably made this thread too long :) Happy to discuss more over zulip etc.

Reply:
> so, you're essentially just saying that "label" is the identifier? that feels icky.

Even though the thread might be too long :joy:, I just want to add my strong :+1: here, I instantly wanted to go in and change the id in `another-plugin` to `my-plugin.mysubmenu` so that it would match where it was going. If two labels are the same, but the ids are different, I would expect two different labels, perhaps with some note to disambiguate them.

Reply:
> I instantly wanted to go in and change the id in another-plugin to my-plugin.mysubmenu so that it would match where it was going. If two labels are the same, but the ids are different, I would expect two different labels, perhaps with some note to disambiguate them.

could we ever get in a situation though then where a plugin references a menu location that doesn't exist? say if the other plugin where that location was originally defined wasn't installed? if we can always figure out something sensible to do in that case then i'd feel much better about this (which is what i liked about nathan's approach)

Reply:
> could we ever get in a situation though then where a plugin references a menu location that doesn't exist? say if the other plugin where that location was originally defined wasn't installed?

absolutely. so we'll need to handle a missing menu id gracefully.  I'll look into how vscode does it